endpoints/events/vtrack.go

```diff
@@ -0,0 +1,312 @@
+
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(d)
+	}
+}
+
+/**
+ * Create vast url tracking
+ */
+func GetVastUrlTracking(externalUrl string, bidid string, bidder string, accountId string, timestamp int64) string {
+
+	eventReq := &EventRequest{
+		Type:      Imp,
+		Bidid:     bidid,
+		AccountID: accountId,
+		Bidder:    bidder,
+		Timestamp: timestamp,
+		Format:    Blank,
+	}
+
+	return EventRequestToUrl(externalUrl, eventReq)
+}
+
+/**
+ * Parses a BidCacheRequest from an HTTP Request
+ */
+func (v *vtrackEndpoint) parseVTrackRequest(httpRequest *http.Request) (req *BidCacheRequest, err error) {
+	req = &BidCacheRequest{}
+	err = nil
+
+	// Pull the request body into a buffer, so we have it for later usage.
+	lr := &io.LimitedReader{
+		R: httpRequest.Body,
+		N: v.Cfg.MaxRequestSize + 1,
+	}
+
+	defer httpRequest.Body.Close()
+	requestJson, err := ioutil.ReadAll(lr)
+	if err != nil {
+		return req, err
+	}
+
+	// Check if the request size was too large
+	if lr.N <= 0 {
+		err = fmt.Errorf("request size exceeded max size of %d bytes", v.Cfg.MaxRequestSize)
+		return req, err
+	}
+
+	if len(requestJson) == 0 {
+		err = fmt.Errorf("request body is empty")
+		return req, err
+	}
+
+	if err := json.Unmarshal(requestJson, req); err != nil {
+		return req, err
+	}
+
+	for _, bcr := range req.Puts {
+		if bcr.BidID == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidid' is required field and can't be empty")})
+			return req, err
+		}
+
+		if bcr.Bidder == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidder' is required field and can't be empty")})
+			return req, err
+		}
+	}
+
+	return req, nil
+}
+
+/**
+ * Cache BidCacheRequest data
+ */
+func (v *vtrackEndpoint) cachePutObjects(req *BidCacheRequest, biddersAllowingVastUpdate []string, accountId string, timeout int64) ([]string, []error) {
+	var cacheables []prebid_cache_client.Cacheable
+	sort.Strings(biddersAllowingVastUpdate)
+
+	for _, c := range req.Puts {
+
+		nc := &prebid_cache_client.Cacheable{
+			Type:       c.Type,
+			Data:       c.Data,
+			TTLSeconds: c.TTLSeconds,
+			Key:        c.Key,
+		}
+
+		if contains(biddersAllowingVastUpdate, c.Bidder) && nc.Data != nil {
+			nc.Data = modifyVastXml(v.Cfg.ExternalURL, nc.Data, c.BidID, c.Bidder, accountId, c.Timestamp)
+		}
+
+		cacheables = append(cacheables, *nc)
+	}
+
+	t := time.Now().Add(time.Duration(timeout) * time.Millisecond)
+	ctx, cancel := context.WithDeadline(context.Background(), t)
+	defer cancel()
+
+	return v.Cache.PutJson(ctx, cacheables)
+}
+
+/**
+ * Returns list of bidders that allow VAST XML modification.
+ */
+func biddersAllowingVastUpdate(req *BidCacheRequest, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) []string {
+	var bl = make(map[string]struct{})
+
+	for _, bcr := range req.Puts {
+		if _, ok := bl[bcr.Bidder]; isAllowVastForBidder(&bcr, bidderInfos, allowUnknownBidder) && !ok {
+			bl[bcr.Bidder] = struct{}{}
+		}
+	}
+
+	l := len(bl)
```

Comment:
Please consider using a different name for `l`. It looks too much like a `1` IMHO and made line 252 hard to read / understand.